felix/wireguard/bootstrap.go

```diff
@@ -0,0 +1,122 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/projectcalico/calico/felix/config"
+	"github.com/projectcalico/calico/felix/netlinkshim"
+	"github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	cerrors "github.com/projectcalico/calico/libcalico-go/lib/errors"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+)
+
+// BootstrapHostConnectivity forces WireGuard peers with hostencryption enabled to communicate with this node unencrypted.
+// This ensures connectivity in scenarios where we have lost our WireGuard config, but will be sent WireGuard traffic
+// e.g. after a node restart, during felix startup, when we need to fetch config from Typha (calico/issues/5125)
+func BootstrapHostConnectivity(configParams *config.Config, getWireguardHandle func() (netlinkshim.Wireguard, error), calicoClient clientv3.Interface) error {
+	wgDeviceName := configParams.WireguardInterfaceName
+	nodeName := configParams.FelixHostname
+
+	if !configParams.WireguardHostEncryptionEnabled {
+		return nil
+	}
+
+	logCtx := log.WithFields(log.Fields{
+		"iface":    wgDeviceName,
+		"hostName": nodeName,
+		"ref":      "wgBootstrap",
+	})
+
+	logCtx.Debug("Bootstrapping wireguard")
+
+	var storedPublicKey string
+	var kernelPublicKey string
+	const (
+		backoffDuration  = 2 * time.Second
+		backoffExpFactor = 2
+		backoffMax       = 32 * time.Second
+		jitter           = 0.2
+	)
+	maxRetries := 3
+	expBackoffMgr := wait.NewExponentialBackoffManager(
+		backoffDuration, backoffMax, time.Minute, backoffExpFactor, jitter, clock.RealClock{})
+	defer expBackoffMgr.Backoff().Stop()
+
+	wg, err := getWireguardHandle()
+	if err != nil {
+		logCtx.Info("Couldn't acquire WireGuard handle, treating public key as unset")
+	} else {
+		kernelPublicKey = getPublicKey(logCtx, wgDeviceName, wg).String()
+		defer wg.Close()
+	}
+
+	// make a few attempts to read our publickey from the datastore, compare, and update if required
+	for r := 0; r < maxRetries; r++ {
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		thisNode, err := calicoClient.Nodes().Get(ctx, nodeName, options.GetOptions{})
+		cancel()
+		if err != nil {
+			logCtx.WithError(err).Debug("Couldn't fetch node config from datastore, retrying")
+			<-expBackoffMgr.Backoff().C() // safe to block here as we're not dependent on other threads
+			continue
+		}
+
+		// if there is any config mismatch, wipe the datastore's publickey (forces peers to send unencrypted traffic)
+		storedPublicKey = thisNode.Status.WireguardPublicKey
+		if storedPublicKey != kernelPublicKey {
+			logCtx.Debug("Found mismatch between kernel and datastore WireGuard keys. Clearing stale key from datastore")
```

Comment:
Boost to `Info`, should only see this log once from Felix so no need to be "quiet"